doc/cpd/101.rst

```diff
@@ -11,117 +11,182 @@ Description
+
+
+::
+
+                     +-----------------+
+     Data Request    |                 |
+   +---------------->+ Storage Service |
+                     |                 | +---------------+
+                     +-------+---------+ |               |
+                             |           |   Custodia    |
+                             |           |               |
+                             |           +---------------+
+                             v                 ^
+                     +-------+-------+      Yes|
+                     |               |         |
+                     |    Secret?    +---------+
+                     |               |         |
+                     +---------------+       No|
+                                               |
+                                               |
+                                               v
+                                         +-----------------------+
+                                         |                       |
+                                         |   Storage Handler(s)  |
+                                         |                       |
+                                         +-----------------------+
 
-* all have the same key
-* each instance have a copy of other instances public keys
-* use the bus to request the encryption key if the local key store is empty
 
 
 Additional Libraries
 ~~~~~~~~~~~~~~~~~~~~
-To facilitate easy access to GnuPG, `python-gnupg <http://pythonhosted.org/gnupg/>`_ would
-need to be installed. This library utilizes ``gpg`` directly so it would need to be
-available as well.
+`Custodia <https://custodia.readthedocs.io/en/latest/>`_ would be a required subsystem.
+Custodia would be installed as part of Commissaire.
 
 
 StorageService Updates
 ~~~~~~~~~~~~~~~~~~~~~~
-The StorageService would need to handle decrypting the key back into a base64 encoded
-string before handing it back to the requesting service.
+The StorageService would need to to know when to use Custodia versus the configured
+``StorageHandler``(s). It would look at the ``_secrets`` attribute on the instance and,
+if set to ``True`` would use the secrets handler.
+
+The secrets handler would be automatically added to the ``StorageHandlerManager`` and
+would require no special configuration by the operator.
 
 
 Model Updates
 ~~~~~~~~~~~~~
-A way to denote encrypted data in a model will need to be added. This will allow the
-``StorageService`` to note when a field needs decrypting. This would likely be an
-item set in the ``_attribute_map``.
+Sensitive items would be pulled out from the ``Host`` model into it's own model. For simplicity,
+the model should be named after the REST endpoint that has traditionally returned the data: ``HostCreds``.
+The models would match based on their primary keys: ``address``.
+
+The ``Model`` class would also gain a new attribute to denote the model is secret. A secret model would
+denote ``_secrets = True``.
```

Comment:
Instead of an attribute, I suggest a class hierarchy.
```
       Model
         |
  +------+------+
  |             |
  |        SecretModel
  |             |
 Host       HostCreds
```
If nothing else, it would make for clearer documentation in APIs where we expect a *secret* model as opposed to any type of model.
```
:param secret_model: A model containing sensitive data
:type secret_model: commissaire.models.SecretModel
```